### ISSUE TITLE
Defining a custom postcircumfix operator causes multi-second hang 

### ISSUE THREAD

**Issue by markjreed** [rakudo/rakudo#5715]
## The Problem

I tried defining a custom postcircumfix indexing operator; just processing the definition makes rakudo hang for multiple seconds. 

## Expected Behavior

Compiling a sub shouldn't take multiple seconds.

## Actual Behavior

The larger the file containing the definition, the longer the hang.  Even if none of the code is actually executed. A 35-line file took 10 seconds to compile before executing any code.

## Steps to Reproduce
foo.raku:
```
#!/usr/bin/env raku
say DateTime.now;
exit;

sub postcircumfix:<᚜  ᚛>(@container,*@index) {
    my $c = @container;
    for @index -> $index {
        return (Any) if $index < 0 || !$c[$index];
        $c = $c[$index];
    }
    return $c;
}
```
```sh
$ date; raku foo.raku
Wed Dec  4 09:19:49 EST 2024
2024-12-04T09:19:53.282199-05:00
```

## Environment
* Operating system:  macOS 14.4.1, Ubuntu 24.04.1
* Compiler version (`rakudo -v` or `raku -v`): v2022.12, v2024.10

**Comment by lizmat** [comment]
```
% time raku 27
2024-12-04T16:50:15.424468+01:00
real	1.05s
user	1.05s
sys	0.03s
```
does that look like the amount of time you expected (this was done on an M1)

FWIW, custom operators require a slangification of the current slang, which is expensive as it is still using a brute force approach.

Is the computer you're using particularly slow or in heavy use?

**Comment by markjreed** [comment]
I was using my M2 MacBook Air, which wasn't busy doing anything else at the time.  I get about 1.8 seconds on my 2017 iMac Pro, and 0.8 seconds on my M3 MBP.  

I had started to post on r/rakulang before deciding that a bug report made more sense, and in that post I was going to ask if there was some rewriting going on; I'm guessing that's what  slangification is.

As another test, I tried adding the sub to the top of one of my existing library modules, which is about 100 LoC, and `raku -e 'use MyMod'` took almost 30 seconds.  Of course precompilation does a world of good for that use case. So did moving the sub to the bottom of the file, since I guess that meant the rest of it didn't have to be slangified.

Thanks for the reply.

**Comment by lizmat** [comment]
> I'm guessing that's what slangification is.

Indeed.  Essentially it clones the grammar and then mixes in additional roles.

>  Of course precompilation does a world of good for that use case

Indeed, once precompiled, there is no difference between core operators, and custom operators: they're all just code blocks that are being called with arguments.

> moving the sub to the bottom of the file, since I guess that meant the rest of it didn't have to be slangified.

That behaviour I find interesting.  That feels like we're also deep cloning something that actually may not need to be cloned.
